client/src/main/java/io/pravega/client/stream/impl/EventStreamWriterImpl.java

```diff
@@ -224,12 +224,32 @@ public void flush() {
                         // available writers.
                         success = false;
                         log.warn("Flush on segment {} failed due to {}, it will be retried.", writer.getSegmentName(), e.getMessage());
+                        tryWaitForSuccessors();
                         break;
                     }
                 }
             }
         }
     }
+    
+    /**
+     * This is used by flush to optimistically wait for the `handleLogSealed` work to be completed
+     * to prevent a busy loop while waiting for the segments to be updated following a sealed
+     * segment. Please note there are no guarantees about when this method returns. It can be
+     * thought of as a sleep.
+     */
+    @GuardedBy("writeFlushLock")
+    private void tryWaitForSuccessors() {
+        while (!sealedSegmentQueue.isEmpty()) {
+            // A background thread should be waking up to process things. Give it a moment.
+            Exceptions.handleInterrupted(() -> Thread.sleep(10));
```

Comment:
A clarification question: what would we need to do to wait on the processing of the sealed segment rather than set a fixed amount of time to wait in each loop iteration?

Reply:
This loop is blocking on the `sealedSegmentQueue` becoming empty. Once the queue is empty and the `writeSealLock` is released we should return. However there is no built in way to block on a queue becoming empty. Hence it is polling. We could probably make a custom data structure to provide this capability if we want.